File: algorithms/medium/replace_words.ex
defmodule Solution do
  defmodule Trie do
    defmodule TrieNode do
      defstruct children: %{}, is_end: false
    end

    def new(), do: %TrieNode{}

    def insert(tree, s) when is_binary(s) do
      insert(tree, String.graphemes(s))
    end

    def insert(nil, cs), do: insert(%TrieNode{}, cs)

    def insert(tree, []), do: %{tree | is_end: true}

    def insert(tree, [c | cs]) do
      %{tree | children: Map.put(tree.children, c, insert(tree.children[c], cs))}
    end

    def search(tree, s) do
      search(tree, [], String.graphemes(s))
    end

    def search(nil, _ss, _cs), do: nil
    def search(tree, _ss, []), do: nil

    def search(tree, ss, [c | cs]) do
      cond do
        tree.is_end -> ss |> Enum.reverse() |> Enum.join()
        tree.children[c] -> search(tree.children[c], [c | ss], cs)
        true -> nil
      end
    end
  end

  @spec replace_words(dictionary :: [String.t()], sentence :: String.t()) :: String.t()
  def replace_words(dictionary, sentence) do
    tree =
      dictionary
      |> Enum.reduce(Trie.new(), &Trie.insert(&2, &1))

    sentence
    |> String.split()
    |> Enum.map(fn s -> Trie.search(tree, s) || s end)
    |> Enum.join(" ")
  end
end

File: algorithms/medium/replace_words.rs
use std::collections::HashMap;

struct TrieNode {
    children: HashMap<char, TrieNode>,
    is_end: bool,
}

impl TrieNode {
    pub fn new() -> Self {
        TrieNode{
            children: HashMap::new(),
            is_end: false,
        }
    }
}

struct Trie {
    root: TrieNode,
}

impl Trie {
    fn new() -> Self {
        Trie {
            root: TrieNode::new(),
        }
    }

    fn insert(&mut self, s: &str) {
        let mut curr = &mut self.root;
        for c in s.chars() {
            curr = curr.children.entry(c).or_insert(TrieNode::new());
        }
        curr.is_end = true;
    }

    fn search(&self, s: &str) -> Option<String> {
        let mut curr = &self.root;
        let mut tmp = String::new();
        for c in s.chars() {
            if let Some(v) = curr.children.get(&c) {
                tmp.push(c);
                if v.is_end {
                    return Some(tmp);
                }
                curr = v;
            } else {
                return None;
            }
        }

        None
    }
}

impl Solution {
    pub fn replace_words(dictionary: Vec<String>, sentence: String) -> String {
        let mut t = Trie::new();
        for w in &dictionary {
            t.insert(w);
        }

        let arr: Vec<&str> = sentence.split(' ').collect();
        let mut ans = Vec::new();
        for s in arr {
            ans.push(t.search(s).unwrap_or(s.to_string()))
        }

        ans.join(" ")
    }
}
